livre/core/pipeline/PipeFilter.cpp

```diff
@@ -0,0 +1,235 @@
+
+/* Copyright (c) 2011-2016, EPFL/Blue Brain Project
+ *                     Ahmet Bilgili <ahmet.bilgili@epfl.ch>
+ *
+ * This file is part of Livre <https://github.com/BlueBrain/Livre>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <livre/core/pipeline/InputPort.h>
+#include <livre/core/pipeline/OutputPort.h>
+#include <livre/core/pipeline/PromiseMap.h>
+#include <livre/core/pipeline/FutureMap.h>
+#include <livre/core/pipeline/PipeFilter.h>
+#include <livre/core/pipeline/Future.h>
+#include <livre/core/pipeline/Promise.h>
+#include <livre/core/pipeline/Filter.h>
+
+namespace livre
+{
+
+struct PipeFilter::Impl
+{
+    typedef std::map< std::string, OutputPort > OutputPortMap;
+    typedef std::map< std::string, InputPort > InputPortMap;
+
+    Impl( PipeFilter& pipeFilter,
+                    const std::string& name,
+                    FilterPtr&& filter )
+        : _pipeFilter( pipeFilter )
+        , _name( name )
+        , _filter( std::move( filter ))
+    {
+        for( const DataInfo& dataInfo: _filter->getInputDataInfos( ))
+        {
+            _inputMap.emplace( std::piecewise_construct,
+                               std::forward_as_tuple( dataInfo.first ),
+                               std::forward_as_tuple( dataInfo ));
+        }
+
+        for( const DataInfo& dataInfo: _filter->getOutputDataInfos( ))
+        {
+            _outputMap.emplace( std::piecewise_construct,
+                                std::forward_as_tuple( dataInfo.first ),
+                                std::forward_as_tuple( dataInfo ));
+        }
+    }
+
+    bool hasInputPort( const std::string& portName ) const
+    {
+        return _inputMap.count( portName ) > 0;
+    }
+
+    bool hasOutputPort( const std::string& portName ) const
+    {
+        return _outputMap.count( portName ) > 0;
+    }
+
+    void throwPortError( const std::string& portName ) const
+    {
+        LBTHROW( std::runtime_error( std::string( "There is no port with name: ")
+                                                  + portName ));
+    }
+
+    void execute()
+    {
+        Futures inputFutures;
+        for( const auto& pair: _inputMap )
```

Comment:
Find a better name than pair for all the loops in this file. Something like namePort is more meaningful.